crates/goose-cli/src/utils.rs

```diff
@@ -0,0 +1,54 @@
+/// Utility functions for safe string handling and other common operations
+
+/// Safely truncate a string at character boundaries, not byte boundaries
+///
+/// This function ensures that multi-byte UTF-8 characters (like Japanese, emoji, etc.)
+/// are not split in the middle, which would cause a panic.
+///
+/// # Arguments
+/// * `s` - The string to truncate
+/// * `max_chars` - Maximum number of characters to keep
+///
+/// # Returns
+/// A truncated string with "..." appended if truncation occurred
+pub fn safe_truncate(s: &str, max_chars: usize) -> String {
+    if s.chars().count() <= max_chars {
+        s.to_string()
+    } else {
+        let truncated: String = s.chars().take(max_chars.saturating_sub(3)).collect();
+        format!("{}...", truncated)
+    }
```

Comment:
Totally fine as is, just a note this walks the string twice. Hopefully we aren't passing large file data here anyways.